### ISSUE TITLE
Problem with the RazorAndMarkdownBlog example

### ISSUE THREAD

**Issue by yetanotherchris** [statiqdev/Statiq#78]
I'm trying to run the RazorAndMarkdownBlog example from source but getting this problem. Is there something I'm missing?

```
       Error while executing module ReadFiles: System.AggregateException: One or more errors occurred. ---> System
rgumentException: Document source must be unique within the pipeline.
  at Wyam.Core.Pipelines.Pipeline.AddDocumentSource(String source) in C:\Code\Wyam\Wyam.Core\Pipelines\Pipeline.cs:lin
263
  at Wyam.Core.Documents.Document.Clone(String source, Stream stream, IEnumerable`1 items, Boolean disposeStream) in C
Code\Wyam\Wyam.Core\Documents\Document.cs:line 217
  at Wyam.Core.Modules.ReadFiles.<>c__DisplayClass11_2.<Execute>b__2(String file) in C:\Code\Wyam\Wyam.Core\Modules\Re
Files.cs:line 89
  at System.Linq.Parallel.SelectQueryOperator`2.SelectQueryOperatorEnumerator`1.MoveNext(TOutput& currentElement, TKey
currentKey)
  at System.Linq.Parallel.PipelineSpoolingTask`2.SpoolingWork()
  at System.Linq.Parallel.SpoolingTaskBase.Work()
  at System.Linq.Parallel.QueryTask.BaseWork(Object unused)
  at System.Linq.Parallel.QueryTask.<>c__DisplayClass0.<_cctor>b__1(Object o)
  at System.Threading.Tasks.Task.InnerInvoke()
  at System.Threading.Tasks.Task.Execute()
  --- End of inner exception stack trace ---
  at System.Linq.Parallel.QueryTaskGroupState.QueryEnd(Boolean userInitiatedDispose)
  at System.Linq.Parallel.AsynchronousChannelMergeEnumerator`1.MoveNextSlowPath()
  at System.Linq.Parallel.AsynchronousChannelMergeEnumerator`1.MoveNext()
  at System.Linq.Parallel.QueryOpeningEnumerator`1.MoveNext()
  at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
  at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
  at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
  at System.Collections.Immutable.ImmutableArray.CreateRange[T](IEnumerable`1 items)
  at System.Collections.Immutable.ImmutableArray.ToImmutableArray[TSource](IEnumerable`1 items)
  at Wyam.Core.Pipelines.Pipeline.Execute(ExecutionContext context, IEnumerable`1 modules, ImmutableArray`1 documents)
n C:\Code\Wyam\Wyam.Core\Pipelines\Pipeline.cs:line 143
--> (Inner Exception #0) System.ArgumentException: Document source must be unique within the pipeline.
  at Wyam.Core.Pipelines.Pipeline.AddDocumentSource(String source) in C:\Code\Wyam\Wyam.Core\Pipelines\Pipeline.cs:lin
263
  at Wyam.Core.Documents.Document.Clone(String source, Stream stream, IEnumerable`1 items, Boolean disposeStream) in C
Code\Wyam\Wyam.Core\Documents\Document.cs:line 217
  at Wyam.Core.Modules.ReadFiles.<>c__DisplayClass11_2.<Execute>b__2(String file) in C:\Code\Wyam\Wyam.Core\Modules\Re
Files.cs:line 89
  at System.Linq.Parallel.SelectQueryOperator`2.SelectQueryOperatorEnumerator`1.MoveNext(TOutput& currentElement, TKey
currentKey)
  at System.Linq.Parallel.PipelineSpoolingTask`2.SpoolingWork()
  at System.Linq.Parallel.SpoolingTaskBase.Work()
  at System.Linq.Parallel.QueryTask.BaseWork(Object unused)
  at System.Linq.Parallel.QueryTask.<>c__DisplayClass0.<_cctor>b__1(Object o)
  at System.Threading.Tasks.Task.InnerInvoke()
  at System.Threading.Tasks.Task.Execute()<---

       Executed pipeline "Posts" (1/2) in 181 ms resulting in 0 output document(s)
   Executing pipeline "Home" (2/2) with 3 child module(s)
           1 errors compiling /home.cshtml:
home.cshtml(5,21): error CS0021: Cannot apply indexing with [] to an expression of type 'IDocumentCollection'
           Error while executing module Razor: System.AggregateException: One or more errors occurred. ---> System.Agg
gateException: One or more errors occurred. ---> System.Exception: /home.cshtml(5,21): error CS0021: Cannot apply inde
ng with [] to an expression of type 'IDocumentCollection'
  --- End of inner exception stack trace ---
  at Wyam.Modules.Razor.Microsoft.AspNet.Mvc.Razor.Compilation.RazorCompilationService.Compile(RelativeFileInfo file,
ring compilationContent) in C:\Code\Wyam\Wyam.Modules.Razor\Mvc\Microsoft.AspNet.Mvc.Razor\Compilation\RazorCompilatio
ervice.cs:line 125
  at Wyam.Modules.Razor.Microsoft.AspNet.Mvc.Razor.Compilation.RazorCompilationService.Compile(RelativeFileInfo file)
 C:\Code\Wyam\Wyam.Modules.Razor\Mvc\Microsoft.AspNet.Mvc.Razor\Compilation\RazorCompilationService.cs:line 61
  at Wyam.Modules.Razor.Microsoft.AspNet.Mvc.Razor.VirtualPathRazorPageFactory.CreateInstance(String relativePath, Str
m stream) in C:\Code\Wyam\Wyam.Modules.Razor\Mvc\Microsoft.AspNet.Mvc.Razor\VirtualPathRazorPageFactory.cs:line 75
  at Wyam.Modules.Razor.Microsoft.AspNet.Mvc.Razor.RazorViewEngine.GetRazorPageResult(ViewContext context, String page
me, Boolean isPartial, Stream stream) in C:\Code\Wyam\Wyam.Modules.Razor\Mvc\Microsoft.AspNet.Mvc.Razor\RazorViewEngin
cs:line 130
  at Wyam.Modules.Razor.Microsoft.AspNet.Mvc.Razor.RazorViewEngine.GetView(ViewContext context, String viewName, Strea
stream) in C:\Code\Wyam\Wyam.Modules.Razor\Mvc\Microsoft.AspNet.Mvc.Razor\RazorViewEngine.cs:line 76
  at Wyam.Modules.Razor.Razor.<>c__DisplayClass7_0.<Execute>b__1(IDocument x) in C:\Code\Wyam\Wyam.Modules.Razor\Razor
s:line 69
  at System.Threading.Tasks.Parallel.<>c__DisplayClass21`2.<ForEachWorker>b__22(Int32 i)
  at System.Threading.Tasks.Parallel.<>c__DisplayClass9`1.<ForWorker>b__11()
  at System.Threading.Tasks.Task.InnerInvoke()
  at System.Threading.Tasks.Task.InnerInvokeWithArg(Task childTask)
  at System.Threading.Tasks.Task.<>c__DisplayClass0.<ExecuteSelfReplicating>b__1(Object )
  --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
  at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
  at System.Threading.Tasks.Task.Wait()
  at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions paralle
ptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
  at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IList`1 list, ParallelOptions parallelOptions, Acti
`1 body, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverythi
, Func`1 localInit, Action`1 localFinally)
  at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IEnumerable`1 source, ParallelOptions parallelOptio
, Action`1 body, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWith
erything, Func`1 localInit, Action`1 localFinally)
  at System.Threading.Tasks.Parallel.ForEach[TSource](IEnumerable`1 source, Action`1 body)
  at Wyam.Modules.Razor.Razor.Execute(IReadOnlyList`1 inputs, IExecutionContext context) in C:\Code\Wyam\Wyam.Modules.
zor\Razor.cs:line 59
  at Wyam.Core.Pipelines.Pipeline.Execute(ExecutionContext context, IEnumerable`1 modules, ImmutableArray`1 documents)
n C:\Code\Wyam\Wyam.Core\Pipelines\Pipeline.cs:line 142
--> (Inner Exception #0) System.AggregateException: One or more errors occurred. ---> System.Exception: /home.cshtml(5
1): error CS0021: Cannot apply indexing with [] to an expression of type 'IDocumentCollection'
  --- End of inner exception stack trace ---
  at Wyam.Modules.Razor.Microsoft.AspNet.Mvc.Razor.Compilation.RazorCompilationService.Compile(RelativeFileInfo file,
ring compilationContent) in C:\Code\Wyam\Wyam.Modules.Razor\Mvc\Microsoft.AspNet.Mvc.Razor\Compilation\RazorCompilatio
ervice.cs:line 125
  at Wyam.Modules.Razor.Microsoft.AspNet.Mvc.Razor.Compilation.RazorCompilationService.Compile(RelativeFileInfo file)
 C:\Code\Wyam\Wyam.Modules.Razor\Mvc\Microsoft.AspNet.Mvc.Razor\Compilation\RazorCompilationService.cs:line 61
  at Wyam.Modules.Razor.Microsoft.AspNet.Mvc.Razor.VirtualPathRazorPageFactory.CreateInstance(String relativePath, Str
m stream) in C:\Code\Wyam\Wyam.Modules.Razor\Mvc\Microsoft.AspNet.Mvc.Razor\VirtualPathRazorPageFactory.cs:line 75
  at Wyam.Modules.Razor.Microsoft.AspNet.Mvc.Razor.RazorViewEngine.GetRazorPageResult(ViewContext context, String page
me, Boolean isPartial, Stream stream) in C:\Code\Wyam\Wyam.Modules.Razor\Mvc\Microsoft.AspNet.Mvc.Razor\RazorViewEngin
cs:line 130
  at Wyam.Modules.Razor.Microsoft.AspNet.Mvc.Razor.RazorViewEngine.GetView(ViewContext context, String viewName, Strea
stream) in C:\Code\Wyam\Wyam.Modules.Razor\Mvc\Microsoft.AspNet.Mvc.Razor\RazorViewEngine.cs:line 76
  at Wyam.Modules.Razor.Razor.<>c__DisplayClass7_0.<Execute>b__1(IDocument x) in C:\Code\Wyam\Wyam.Modules.Razor\Razor
s:line 69
  at System.Threading.Tasks.Parallel.<>c__DisplayClass21`2.<ForEachWorker>b__22(Int32 i)
  at System.Threading.Tasks.Parallel.<>c__DisplayClass9`1.<ForWorker>b__11()
  at System.Threading.Tasks.Task.InnerInvoke()
  at System.Threading.Tasks.Task.InnerInvokeWithArg(Task childTask)
  at System.Threading.Tasks.Task.<>c__DisplayClass0.<ExecuteSelfReplicating>b__1(Object )
--> (Inner Exception #0) System.Exception: /home.cshtml(5,21): error CS0021: Cannot apply indexing with [] to an expre
ion of type 'IDocumentCollection'<---

---
```

**Comment by yetanotherchris** [comment]
If it's any help, I refactored the Examples Tests to this, and all of them seem to be failing:

```
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using NUnit.Framework;
using Wyam.Core;

namespace Wyam.Examples.Tests
{
    [TestFixture(Category = "ExcludeFromAppVeyor")]
    public class ExamplesFixture
    {
        private static IEnumerable<string> _paths;

        public static IEnumerable<string> Paths
        {
            get
            {
                if (_paths == null)
                {
                    string rootPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Examples");
                    while (!Directory.Exists(rootPath))
                    {
                        rootPath = Directory.GetParent(rootPath).Parent.FullName;
                        rootPath = Path.Combine(rootPath, "Examples");
                    }

                    _paths = Directory.EnumerateDirectories(rootPath);
                }

                return _paths;
            }
        }

        [Test]
        [TestCaseSource(typeof(ExamplesFixture), "Paths")]
        public void ExecuteAllExamples(string example)
        {
            Console.WriteLine("Executing example " + example);
            Engine engine = new Engine
            {
                RootFolder = example
            };
            string config = Path.Combine(example, "config.wyam");
            if (File.Exists(config))
            {
                Console.WriteLine("Loading configuration file");
                engine.Configure(File.ReadAllText(config));
            }
            engine.Execute();
        }
    }
}
```

**Comment by daveaglick** [comment]
Thanks for checking the project out. First off, I love the test refactoring! That makes it so much easier to see which example is failing. That said, I'm having trouble reproducing the exception you're seeing:
![2015-10-13_14h11_03](https://cloud.githubusercontent.com/assets/1020407/10463767/55b23f9c-71b4-11e5-88a2-b5ff49877350.png)

A couple questions:
- Are you working with the `master` branch or the `develop` branch?
- Have you made any changes to the example code (specifically the `config.wyam` files) or are they straight from the repo?
- What happens if you run the failing example via the command line application by building Wyam (in Debug) and then running this from the example directory: `e:\Code\Wyam\Examples\RazorAndMarkdownBlog>..\..\Wyam\bin\Debug\Wyam.exe --verbose`

Though it's probably not too relevant yet, here's some background on the actual error "Document source must be unique within the pipeline." Every document gets a source property which tells the system where the document came from. It could be the full path to a file, a URL, etc. The error is produced when a pipeline loads a given source more than once. There are a couple reasons why it's important that this produce an error:
- The caching mechanism partly relies on the source and having multiple documents with the same source would confuse it (the system might get the wrong cache information).
- It's used for reporting errors and warning for a specific document and if there were more than one document with the same source these errors and warnings would be ambiguous.
- It suggests an inefficiency or mistake in the config file. A given source should only be loaded once for performance reasons, and there are modules to help reuse it if needed.

I don't think any of these are at issue here though. The examples should "just work" so I suspect something else is going on...

**Comment by yetanotherchris** [comment]
I've switched to master as I was on develop, but the problem is the same in both the console and unit tests.

The tests were failing as the path (in the source I posted) was pointing to the examples directory, however the Configurator is expecting it to be run in the bin folder, so it can pickup the various Wyam.Module assemblies (if I've understood correctly, via the line `AppDomain.CurrentDomain.AssemblyResolve += ResolveAssembly;`).

I got around this by adding a dirty hack into the Configurator constructor:

```
foreach (string file in Directory.EnumerateFiles(AppDomain.CurrentDomain.BaseDirectory, "*.dll"))
{
    AddAssembly(Assembly.LoadFile(file));
}
```

The tests work ok for that, but on the console I get the same exception and stacktrace that I posted above.

**Comment by yetanotherchris** [comment]
It seems to be fine for Wyam.Web but Inside `AddDocumentSource` it's trying to re-add post.cshtml to the Hashset for that particular example. I can use Wyam.Web to work from though.

**Comment by daveaglick** [comment]
Interesting... Thanks for helping look at this.

This does indeed look to be some kind of path problem. Another clue is this line in the error output from above: `Executed pipeline "Posts" (1/2) in 181 ms resulting in 0 output document(s)` - it should be finding two Markdown files, `a.md` and `b.md`, but instead it's not finding any. Then when we get to the later pipeline, we crash because the collection of post documents is empty (and since this is a simple example, I haven't added error checking to the template).

The paths are usually driven by the `Engine.RootFolder` which is initially set to `Environment.CurrentDirectory` unless explicitly changed by the configuration file. `Path.Combine()` calls are used to derived the default input and output folders from that.

I'm wondering if it's trying to add the duplicate because it's one level higher than it should be. The root path should usually be set to the same folder where the config file is (the parent of the input and output folders).

I've just committed a couple changes to the develop branch with some extra logging. Do you mind running one more time from the command line with the new changes and the `--verbose` flag and then pasting the output (or maybe attaching since it'll be big)? You can log to a file with `--log log.txt`.

**Comment by daveaglick** [comment]
BTW, the `Configurator` was another good place to look. In the `Configurator.GetAssemblies()` method, it tries to locate all the module assemblies with the following call:

```
Directory.GetFiles(Path.GetDirectoryName(typeof(Configurator).Assembly.Location), "*.dll", SearchOption.AllDirectories)
```

This should start with the same folder as the `Wam.Core` assembly (since that's where `typeof(Configurator).Assembly.Location` points to) and look at all nested paths for .dll files. The fact you're having trouble locating the module assemblies as well is odd.

The verbose output should help here too because it'll tell us exactly where it's looking for assemblies.

**Comment by yetanotherchris** [comment]
Github attach isn't working, or 47k is too much for it so I've pasted it here:

http://pastebin.com/2AFXyC5i

**Comment by daveaglick** [comment]
Got it - that's tremendously helpful. Looks like you were right - it's trying to read `C:\Code\Wyam\Examples\RazorAndMarkdownBlog\Input\posts\post.cshtml` twice in the same `ReadFiles` module, which doesn't make any kind of sense to me. I guess the good news is that the duplicate source check is doing exactly what it should be and stopping/crashing an invalid build. I've got to head out for now, but I'll take a closer look tonight to see if I can repro this.

Thanks a ton for the help.

**Comment by daveaglick** [comment]
I've been able to reproduce the problem. It does appear as though behavior is different for the test runner and console (which isn't the real problem, but also needs to be addressed). Should have fixes for both soon.

**Comment by daveaglick** [comment]
This is now resolved in the develop branch. It turns out there were a couple of problems, the main one being that since Wyam is designed to try and recover from errors by continuing the next pipeline on failure, the example test fixture wasn't asserting that the examples didn't fail. This example has probably been failing for a while and I never realized it.

I also fixed the underlying cause of the failures (there were two) triggered by the way document source was being handled and a previous change to the `IDocumentCollection` interface.

Note to self - this bug exposed a logic problem with the way `IDocument.Source` is used. From now on, once a source is set, cloning the document maintains the original source (which makes sense to me anyway). If you want an entirely new document with a new source, use `IExecutionContext.GetNewDocument()`.

These fixes will go out with the next version, but that's probably a couple weeks off - I'm in the middle of a spike for #70 right now. I'd be happy to build an intermediate distribution for you if needed, but it looks like you're able to build from the develop branch, so that's your best bet.

Thanks again for the report!
